File: python/examples/demo_const_known.py
"""
Demo: smoothing with known parameters.
"""

# simksmoother --- a simultaneous Kalman smoother in sparse linear algebra form.
# Copyright (C) 2026 Mico Mrkaic.
#
# Produced under the guidance and direction of Mico Mrkaic, with the
# assistance of AI (Claude, Anthropic).
#
# This file is part of the simksmoother package.
#
# simksmoother is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the LICENSE file in the package root,
# or <https://www.gnu.org/licenses/>.
#

import numpy as np
from _plot_helpers import plt, save_and_close
import simksmoother as sks

def main(seed=1):
    rng = np.random.default_rng(seed)

    n = 2
    T = 100
    F = np.array([[0.75, 0.15], [0.0, 0.5]])
    H = np.array([[1.0, 0.3], [0.2, 1.0]])
    Q = 0.1 * np.eye(2)
    R = 0.1 * np.eye(2)
    a0 = np.zeros(2)
    P0 = np.eye(2)

    Y, X = sks.simulate_const(T, F, H, Q, R, a0, P0, rng=rng)
    Xhat, J, h, info = sks.smooth_const(Y, F, H, Q, R, a0, P0, return_cov=True)

    print("Known-parameter simultaneous smoother demo")
    print(f"State dimension n = {n}, observations T = {T}")
    print(f"Sparse precision size: {J.shape[0]} x {J.shape[1]}, nnz = {info['nnz']}")

    ll, _ = sks.marginal_loglik(Y, F, H, Q, R, a0, P0)
    print(f"Marginal log-likelihood = {ll:.4f}")

    rmse1 = float(np.sqrt(np.mean((Xhat[0, 1:] - X[0, 1:])**2)))
    rmse2 = float(np.sqrt(np.mean((Xhat[1, 1:] - X[1, 1:])**2)))
    print(f"RMSE state 1 = {rmse1:.4f}")
    print(f"RMSE state 2 = {rmse2:.4f}")

    # Empirical 95% coverage on state 1
    sd = np.array([np.sqrt(info["Ptt"][t][0, 0]) for t in range(1, T + 1)])
    z = (Xhat[0, 1:] - X[0, 1:]) / sd
    coverage = float(np.mean(np.abs(z) < 1.96))
    print(f"Empirical 95% coverage for state 1 = {coverage:.3f}")

    # --- Plot ----------------------------------------------------------
    # State trajectories with 95% posterior bands, plus observation overlay
    # on state 1 (it's the linearly-mostly-loaded one in this H).
    tt = np.arange(T + 1)
    sd1 = np.array([np.sqrt(info["Ptt"][t][0, 0]) for t in range(T + 1)])
    sd2 = np.array([np.sqrt(info["Ptt"][t][1, 1]) for t in range(T + 1)])

    fig, axes = plt.subplots(2, 1, figsize=(9, 6), sharex=True)
    for ax, idx, sd in [(axes[0], 0, sd1), (axes[1], 1, sd2)]:
        ax.fill_between(tt, Xhat[idx] - 1.96 * sd, Xhat[idx] + 1.96 * sd,
                        color="#d8e2f3", label="95% band")
        ax.plot(tt, X[idx], "k-", lw=1.2, label="true")
        ax.plot(tt, Xhat[idx], "r--", lw=1.2, label="smoothed")
        ax.set_ylabel(f"x_{idx+1}")
        ax.legend(loc="best", fontsize=9)
        ax.grid(alpha=0.3)
    # Overlay observation on state 1 panel (using y_1 since H has 1.0 on x_1).
    axes[0].plot(np.arange(1, T + 1), Y[0], "o", ms=3,
                 color="0.6", label="obs (y_1)")
    axes[0].legend(loc="best", fontsize=9)
    axes[0].set_title("Smoothed states with 95% posterior bands")
    axes[-1].set_xlabel("t")
    fig.tight_layout()
    save_and_close(fig, "demo_const_known")

if __name__ == "__main__":
    main()

File: octave/examples/demo_const_known.m
% simksmoother --- a simultaneous Kalman smoother in sparse linear algebra form.
% Copyright (C) 2026 Mico Mrkaic.
%
% Produced under the guidance and direction of Mico Mrkaic, with the
% assistance of AI (Claude, Anthropic).
%
% This file is part of the simksmoother package.
%
% simksmoother is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program; see the LICENSE file in the package root,
% or <https://www.gnu.org/licenses/>.
%

% DEMO_CONST_KNOWN  Known-parameter simultaneous Kalman smoothing.
% (script: cleared and cleared screen removed for demo_main compatibility)
% Add the simksmoother/ source directory to the path, relative to where
% this demo file lives.  Works regardless of the user's CWD.
demo_dir = fileparts(mfilename('fullpath'));
oct_dir  = fullfile(demo_dir, '..', 'simksmoother');
if isfolder(oct_dir), addpath(oct_dir); end

randn('seed', 1);

T  = 100;
F  = [0.8 0.2; 0.0 0.6];
H  = [1.0 0.0];
Q  = [0.10 0.02; 0.02 0.08];
R  = 0.25;
a0 = [0;0];
P0 = eye(2);

[Y, Xtrue] = simks_simulate_const(T, F, H, Q, R, a0, P0);

% Smoother with posterior covariance blocks via selected inversion.
opts.return_cov = true;
[Xhat, J, h, info] = simks_smooth_const(Y, F, H, Q, R, a0, P0, opts);

% Marginal log-likelihood at the true parameters.
ll = simks_marginal_loglik(Y, F, H, Q, R, a0, P0);

fprintf('Known-parameter simultaneous smoother demo\n');
fprintf('State dimension n = %d, observations T = %d\n', size(F, 1), T);
fprintf('Sparse precision size: %d x %d, nnz = %d\n', size(J, 1), size(J, 2), info.nnz);
fprintf('Marginal log-likelihood = %.4f\n', ll);
fprintf('RMSE state 1 = %.4f\n', sqrt(mean((Xhat(1,:) - Xtrue(1,:)).^2)));
fprintf('RMSE state 2 = %.4f\n', sqrt(mean((Xhat(2,:) - Xtrue(2,:)).^2)));

% Empirical coverage of 95% smoothed bands for state 1.
band = zeros(1, T+1);
for t = 0:T
  band(t+1) = 1.96 * sqrt(info.Ptt{t+1}(1,1));
end
cov_state1 = mean(abs(Xhat(1,:) - Xtrue(1,:)) <= band);
fprintf('Empirical 95%% coverage for state 1 = %.3f\n', cov_state1);

% --- Plot ----------------------------------------------------------------
% State trajectories: true, smoothed, +/-2 sigma bands.  State 1 also has
% the noisy observation overlaid (state 2 is unobserved).
band2 = zeros(1, T+1);
for t = 0:T
  band2(t+1) = 1.96 * sqrt(info.Ptt{t+1}(2,2));
end
tt = 0:T;
fig = figure('visible', 'on', 'position', [100 100 900 600]);

subplot(2,1,1);
hold on; box on;
fill([tt, fliplr(tt)], [Xhat(1,:)-band, fliplr(Xhat(1,:)+band)], ...
     [0.85 0.90 0.98], 'edgecolor', 'none');
plot(1:T, Y, 'o', 'color', [0.6 0.6 0.6], 'markersize', 3);
plot(tt, Xtrue(1,:), 'k-',  'linewidth', 1.2);
plot(tt, Xhat (1,:), 'r--', 'linewidth', 1.2);
legend({'95% band', 'observation', 'true', 'smoothed'}, 'location', 'best');
title('State 1: true (observed) coordinate');
xlabel('t'); ylabel('x_1');

subplot(2,1,2);
hold on; box on;
fill([tt, fliplr(tt)], [Xhat(2,:)-band2, fliplr(Xhat(2,:)+band2)], ...
     [0.85 0.90 0.98], 'edgecolor', 'none');
plot(tt, Xtrue(2,:), 'k-',  'linewidth', 1.2);
plot(tt, Xhat (2,:), 'r--', 'linewidth', 1.2);
legend({'95% band', 'true', 'smoothed'}, 'location', 'best');
title('State 2: unobserved coordinate');
xlabel('t'); ylabel('x_2');

fig_dir = fullfile(fileparts(mfilename('fullpath')), 'figures');
if ~exist(fig_dir, 'dir'), mkdir(fig_dir); end
fig_path = fullfile(fig_dir, 'demo_const_known.png');
print(fig, fig_path, '-dpng', '-r120');
drawnow;
fprintf('Figure saved: %s\n', fig_path);
